allow to send custom errors

caub opened the pull request:
This allows to send errors in a more practical way, coupling status and message together

```js
// utils
class ApiError extends Error {
    constructor(status = 500, message = 'Something went wrong') {
        super(message);
        this.status = status;
    }
    toJSON() {
        return { text: this.message };
    }
}
const error = {
    notFound: (s = 'resource') => new ApiError(404, `${s} not found`),
    missingParams: (s = '') => new ApiError(400, `Missing parameters${s ? ': ' + s : ''}`),
    forbidden: () => new ApiError(403, 'Access forbidden'),
};

// routers
app.get('/foo', (req, res) => {
  if (typeof req.body.data !== 'object') {
    return res.send(error.missingParams('data'));
  }
  // ...
})
```
Of course it's possible to use `next(error.missingParams('data'))` and handle the logic in the final error handler middleware, but I think this little addition is helpful, for more complex cases than https://expressjs.com/en/api.html#res.sendStatus

This PR is backward compatible

~~There's just an issue with eslint which can't parse `class`, is it ok to update eslint? the version used here is very old~~

caub committed:
allow to send custom errors status

dougwilson commented:
Hi @caub thanks for your pull request! I think this is a good idea! I do have a couple comments / questions, though:

1. We are trying to remove all the overloads from `res.send`, which is why there is now `res.json`, `res.sendStatus`, etc. Can this logic be added as a `res.sendError` instead, for example?
2. It feels to me that if someone is sending an error, the status code should default to 500, and not 200. This is because 500 is the generic code for a server error, and I'm not sure why someone would want to send an error object but it end up as a 200?
3. May want to add a guard somewhere, or at least acknowledge when the `status` property of the supplied object is something like a string, object, a number like 20000, etc. The answer can be this is an error and it will throw, but we should at least make sure what happens here is a conscious decision. The current implementation will end up varying by Node.js version, as some will typecast `res.statusCode`, even to the number `0`, in these cases.

As an addition to note (1), that would make this backwards-compatible. As it is currently, it's not actually backwards compatible, as users can currently be passing in an object to `res.send` that has a `status` property they are only expecting to be in the resulting JSON output, and not actually get set as the status code. Putting this in a separate method like `res.sendError` would avoid this issue, of course, and then it can land in 4.x.

Also, just as an additional curiosity (and also so we can help write the appropriate documentation around this and answer questions on this), can you explain what cases folks should use this method over `next(error)`, and also when they should use `next(error)` over this method?

Thank you!

caub commented:
Thanks, good ideas,  I thought about a separate `.sendError` initially, but thought it would be simpler for users to just use `.send`

I've modified the PR. I've added `.jsonError` (it's maybe worth adding the `.sendError` alternative),

dougwilson commented:
I know this has sat for a long time, and I apologize for that. Does anyone have any thoughts on this proposal?

Looking at the new changes (once again, sorry for not doing so for so long), I'm struggling to see what the additional API surface is really giving, as it is just under the hood a short hand for like `res.status(err.status).json(err)` (simplified here of course). I feel the downside to this simplification is that is locks in two things: the properties that can contain a status code and that the response is JSON.

caub commented:
Yes you're right, after looking back too, we can close
